### PR TITLE
Added Sudoku example

### PR DESCRIPTION
**Pull request by satyaki3794**
I have added an example for sudoku solver to hpx/examples. The initial state of the board is fixed and the solver uses brute force/backtracking to solve the puzzle or reports if no solution is possible.

### CODE REVIEW

File: examples/sudoku/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2016 Satyaki Upadhyay
+#
+# Distributed under the Boost Software License, Version 1.0. (See accompanying
+# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+
+add_hpx_component(sudoku
+  SOURCE_GLOB "sudoku.cp*"
+  HEADER_GLOB "sudoku.hp*"
```
Comment (zao) [inline examples/sudoku/CMakeLists.txt]: Any particular reasons for the use of globs here, or is it inspired by an example?

Comment (satyaki3794) [inline examples/sudoku/CMakeLists.txt]: Inspired by nqueens example.

File: examples/sudoku/server/sudoku.hpp [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+{
+    typedef std::vector<std::size_t> list_type;
+
+namespace server
+{
+    class board
+        : public hpx::components::component_base<board>
+    {
+    private:
+        list_type list_;
```
Comment (hkaiser) [inline examples/sudoku/server/sudoku.hpp]: Why calling it a 'list' if it's anything but one?

File: examples/sudoku/server/sudoku.hpp [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+#include <hpx/include/async.hpp>
+#include <hpx/include/util.hpp>
+#include <hpx/hpx_fwd.hpp>
+#include <hpx/include/components.hpp>
+#include <hpx/runtime/applier/applier.hpp>
+#include <hpx/runtime/threads/thread_data.hpp>
+#include <hpx/runtime/components/component_type.hpp>
+#include <hpx/runtime/actions/component_action.hpp>
+#include <hpx/runtime/serialization/serialize.hpp>
```
Comment (hkaiser) [inline examples/sudoku/server/sudoku.hpp]: Please try to include headers from `hpx/include` instead, if possible. We'd like to slowly move away from exposing the internal headers.

File: examples/sudoku/server/sudoku.hpp [hunk 4]
```diff
@@ -0,0 +1,216 @@
+//  Copyright (c) 2016 Satyaki Upadhyay
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#if !defined(HPX_SUDOKU_EXAMPLE_SERVER)
+#define HPX_SUDOKU_EXAMPLE_SERVER
+
+#include <iostream>
```
Comment (hkaiser) [inline examples/sudoku/server/sudoku.hpp]: Is including this header really necessary?

File: examples/sudoku/server/sudoku.hpp (hunk 2 above)
Comment (satyaki3794) [inline examples/sudoku/server/sudoku.hpp]: The 9X9 board is represented as a list of 81 elements. Position (i, j) on the board corresponds to (9*i+j) in the list (0-indexed).

File: examples/sudoku/sudoku_client.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
+//  Copyright (c) 2016 Satyaki Upadhyay
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+
+#include <iostream>
```
Comment (hkaiser) [inline examples/sudoku/sudoku_client.cpp]: Would it be viable to use `hpx::cout` instead of the standard streams?

**Comment by hkaiser** [comment]
Overall, very nice. I'd prefer to move the initialization of the board into the main program, though. The way you implemented it the component can be used to solve only a single puzzle. It would be nice it solved arbitrary ones.

**Comment by satyaki3794** [comment]
The initialization function throws a 'future has no valid shared state' exception, similar to nqueens one (please refer here: https://github.com/STEllAR-GROUP/hpx/issues/802#issuecomment-194255571). I tried fixing it, but was unable to.

File: examples/sudoku/server/sudoku.hpp (hunk 2 above)
Comment (hkaiser) [inline examples/sudoku/server/sudoku.hpp]: But you're storing it as a `std::vector`, not a `std::list`, aren't you? Also, a `std::vector<boost::uint8_t>` should be sufficient for your use case (you're storing numbers in the range [0,81] only).

Comment (satyaki3794) [inline examples/sudoku/server/sudoku.hpp]: Okay. I will change the name of the attribute and data_type to uint8_t. Also, it stores numbers in the range [0,9].

File: examples/sudoku/server/sudoku.hpp (hunk 4 above)
Comment (satyaki3794) [inline examples/sudoku/server/sudoku.hpp]: No. I guess I forgot to remove it. I'll make the changes.

**Comment by hkaiser** [comment]
Several general comments:

- Where is the parallelism? Wasn't the plan to solve the puzzle using some parallelization?
- Please get rid of the stubs, we deprecated those (just move the async calls directly into the client)
- The current code throws an exception saying that `this future has no valid shared state: HPX(no_state)`, this is because the code does not actually initialize the board (it doesn't create an instance of that component). Either do

        sudoku::board new_board = hpx::new_<sudoku::board>(locality_);

  or do somethiing equivalent in the constructor of the client object.
- As said before, I'd prefer for the component not to initialize the board with some concrete puzzle, I think this should be done from the application via the constructor of the puzzle object.
- The 'solver' should be part of the puzzle component as well and should be exposed through a `solve` action or somesuch. In the end, the application would look like

        sudoku::board_data data = { ... };
        sudoku::board board = hpx::new_<sudoku::board>(hpx::find_here(), board_data);
        hpx::future<sudoku::board_data> solution = board.solve();

        for (sudoku::cell const& c : data)
            ...

  (or similar).

**Comment by satyaki3794** [comment]
I have implemented the suggested changes. Removed the deprecated stubs and handled the future_has_no_shared_state_exception. Added the feature of taking the starting board configuration from the user along with a default which is provided. Changed the implementation logic of solve_board(). It now returns a future containing the final board configuration, and uses hpx::async() to create new threads asynchronously, and cancellation tokens to signal them to stop, if the parent thread has already found a solution.

**Comment by hkaiser** [comment]
@satyaki3794 Thanks for all of your work on this and sorry for the long delay in responding.

Everything looks fine now except one thing which you might want to consider chaging. Currently the code which asks the user for an initial board layout sits in the component. I think this belongs into the client executable instead. The component is really just the solver, it gets initialized from an initial board layout (wherever that comes from) and returns a solution.

**Comment by satyaki3794** [comment]
Do you think the whole functionality should be in the client executable? Or can I take the input via the client and pass a vector to the action which initializes the board (as opposed to doing everything inside the component action), in which case the component action simply assigns the board config to the one passed as an argument?

**Comment by sithhell** [comment]
What's left to do here?

**Comment by hkaiser** [comment]
> What's left to do here?

See my comment above:

> Everything looks fine now except one thing which you might want to consider changing. Currently the code which asks the user for an initial board layout sits in the component. I think this belongs into the client executable instead. The component is really just the solver, it gets initialized from an initial board layout (wherever that comes from) and returns a solution.

**Comment by hkaiser** [comment]
I'm going to close this without merging. Too many things would have to be changed for this to be a usable example.

**Pull request closed without merging**